actors/builtin/market/market_actor.go

```diff
@@ -427,78 +410,62 @@ func (a Actor) OnMinerSectorsTerminate(rt Runtime, params *OnMinerSectorsTermina
 	return nil
 }
 
-type HandleExpiredDealsParams struct {
-	Deals []abi.DealID // TODO: RLE
-}
+func (a Actor) CronTick(rt Runtime, params *adt.EmptyValue) *adt.EmptyValue {
+	rt.ValidateImmediateCallerIs(builtin.CronActorAddr)
+	amountSlashed := big.Zero()
 
-func (a Actor) HandleExpiredDeals(rt Runtime, params *HandleExpiredDealsParams) *adt.EmptyValue {
-	rt.ValidateImmediateCallerType(builtin.CallerTypesSignable...)
-	var slashed abi.TokenAmount
 	var st State
 	rt.State().Transaction(&st, func() interface{} {
-		slashed = st.updatePendingDealStates(rt, params.Deals, rt.CurrEpoch())
-		return nil
-	})
+		dbe, err := AsSetMultimap(adt.AsStore(rt), st.DealOpsByEpoch)
+		if err != nil {
+			rt.Abortf(exitcode.ErrIllegalState, "failed to load deal opts set: %s", err)
+		}
 
-	// TODO: award some small portion of slashed to caller as incentive
+		updatesNeeded := make(map[abi.ChainEpoch][]abi.DealID)
 
-	_, code := rt.Send(builtin.BurntFundsActorAddr, builtin.MethodSend, nil, slashed)
-	builtin.RequireSuccess(rt, code, "failed to burn funds")
-	return nil
-}
+		for i := st.LastCron; i <= rt.CurrEpoch(); i++ {
```

Comment:
But its not an AMT, so thats really doable

Reply:
Ack, ok.

Not necessarily in this PR, but I'm increasingly thinking we should use an AMT for things indexed by epoch, so that we can traverse in order. The same goes for the cron callback table in the power actor. I think there's decent scope for bugs leading to missed events and junk state left behind otherwise, especially given that cron doesn't necessarily execute on the epoch scheduled (due to null rounds).

In this case, with the values as integers, it could be an AMT of bitfields; the power actor needs to store more metadata tho.